### DISCUSSION TITLE
Testing `getRouter` calls

### DISCUSSION THREAD

**Discussion by jay-tux** [SELab-2/OSOC-2#347] [Polls]
Currently, all route files have a single function called `getRouter`. This function is called to set up an `express.Router` for that route. However, this makes it nigh-impossible to get 100% coverage on the routes files.

There's a few ways to remedy this (some more tiresome than others):
 - Move all `getRouter` calls to a single (other) file and refactor `endpoints.ts` to reflect the changes. Why not move them to `endpoints.ts`? Because that would cause the exact same issue when we're testing `endpoints.ts` - the `getRouter` calls are inside the `endpoints.ts` and would have to be mocked by then (probably).
 - Mock every single utility & database call that happens in any of the routes. This will take an insane amount of time and effort on our part, and it will result in a very, very messy `getRouter.test.ts`.

Another option would be to just say *screw it, we don't need 100% coverage*

**Comment by BramDevlaminck** [comment]
What level of coverage would be able to achieve when we say *screw it*? If this percentage is still 90% or something like that I would think that is still a reasonable goal and it would still allow us to keep the (good) structure we have now.

**Reply by jay-tux** [reply]
For the `router/admin.ts` file, current coverage is
```
---------------------------|---------|----------|---------|---------|-------------------
File                       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
---------------------------|---------|----------|---------|---------|-------------------
  admin.ts                 |   79.48 |      100 |   95.45 |   78.94 | 94-105            
---------------------------|---------|----------|---------|---------|-------------------
```

As this is still yellow (not green), I would improve it

**Reply by BramDevlaminck** [reply]
If this is indeed the coverage we can achieve without changing anything we should change something

**Comment by Mouwrice** [comment]
I fail to understand why you can't test the `getRouter` function? Can you not test that the router returned should only have the attributes added to the router?

**Reply by jay-tux** [reply]
There are no added attributes - express Routers have builtin properties and you pass callbacks to them.
The way I'm trying to test the routers is by mocking the router structure itself and passing mocked requests, to see if the correct functions are called (because equality on functions doesn't make sense).

**Reply by Mouwrice** [reply]
So you are basically testing the express library?

**Comment by Mouwrice** [comment]
So to come back to this. I don't really feel the need the check if the route is correctly setup - because we would partly be testing the framework itself, which is pointless - and we do "integration" tests by calling the routes from the frontend.

Just trick jest by executing the function in a test without doing anything with the result to get to 100% code coverage.

**Reply by jay-tux** [reply]
Sounds like cheating, but with the hours of trying - I'll take it haha

**Reply by Mouwrice** [reply]
Well, yes. That is exactly what it is! 🤣

**Comment by jay-tux** [comment]
Closing with conclusion:

**Testing getRouter would be pointless - as it mostly tests the express framework. True testing can and will be done using begin-to-end tests using SuperTest.**
